File: frontend/src/components/benchmark/content/counting_sort/implementations/counting_sort.cs
using System;
using System.Collections.Generic;

public class CountingSort
{
    private int FindMax(int[] arr)
    {
        int maxVal = arr[0];
        foreach (int val in arr)
        {
            if (val > maxVal) maxVal = val;
        }
        return maxVal;
    }

    private int FindMin(int[] arr)
    {
        int minVal = arr[0];
        foreach (int val in arr)
        {
            if (val < minVal) minVal = val;
        }
        return minVal;
    }

    public void Sort(int[] arr)
    {
        if (arr == null || arr.Length == 0) return;

        int maxVal = FindMax(arr);
        int minVal = FindMin(arr);
        int range = maxVal - minVal + 1;

        int[] count = new int[range];
        int[] output = new int[arr.Length];

        foreach (int val in arr)
        {
            count[val - minVal]++;
        }

        for (int i = 1; i < count.Length; i++)
        {
            count[i] += count[i - 1];
        }

        for (int i = arr.Length - 1; i >= 0; i--)
        {
            output[count[arr[i] - minVal] - 1] = arr[i];
            count[arr[i] - minVal]--;
        }

        for (int i = 0; i < arr.Length; i++)
        {
            arr[i] = output[i];
        }
    }

    public void SortStable(int[] arr)
    {
        Sort(arr);
    }

    public void SortDescending(int[] arr)
    {
        Sort(arr);
        Array.Reverse(arr);
    }
}

File: frontend/src/components/benchmark/content/counting_sort/implementations/counting_sort.go
package main

type CountingSort struct{}

func (cs *CountingSort) findMax(arr []int) int {
    maxVal := arr[0]
    for _, val := range arr {
        if val > maxVal {
            maxVal = val
        }
    }
    return maxVal
}

func (cs *CountingSort) findMin(arr []int) int {
    minVal := arr[0]
    for _, val := range arr {
        if val < minVal {
            minVal = val
        }
    }
    return minVal
}

func (cs *CountingSort) Sort(arr []int) {
    if len(arr) == 0 {
        return
    }

    maxVal := cs.findMax(arr)
    minVal := cs.findMin(arr)
    rangeVal := maxVal - minVal + 1

    count := make([]int, rangeVal)
    output := make([]int, len(arr))

    for _, val := range arr {
        count[val-minVal]++
    }

    for i := 1; i < len(count); i++ {
        count[i] += count[i-1]
    }

    for i := len(arr) - 1; i >= 0; i-- {
        output[count[arr[i]-minVal]-1] = arr[i]
        count[arr[i]-minVal]--
    }

    for i := 0; i < len(arr); i++ {
        arr[i] = output[i]
    }
}

func (cs *CountingSort) SortStable(arr []int) {
    cs.Sort(arr)
}

func (cs *CountingSort) SortDescending(arr []int) {
    cs.Sort(arr)
    for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
        arr[i], arr[j] = arr[j], arr[i]
    }
}